### DISCUSSION TITLE
using-default-interface-methods-for-performance-gains-in-iheaderdictionary/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#408] [Announcements]
# using-default-interface-methods-for-performance-gains-in-iheaderdictionary/

In this post I look at an example of how default interfaces were used to improve the performance of IHeaderDictionary

https://andrewlock.net/using-default-interface-methods-for-performance-gains-in-iheaderdictionary/

**Comment by ADD-David-Antolin** [comment]
Since the concrete classes seem to be overriding all the default interface methods and the performance gains are in the code of the concrete classes, I think the default interface methods in this case have nothing to do with the performance gains. If you quit the default implementations but keep the definitions both HttpRequestHeaders and HttpResponseHeaders still compile and still have the performance gains.

**Reply by andrewlock** [reply]
IMO you're partially right. If the interface was internal, absolutely, you wouldn't need the default interfaces, as they add nothing.

But this is a public interface. People are free to implement the interface in their own concrete types. So making this change without default interface methods would be massively breaking for such a fundamental interface. So it's unlikely the team would have gone for it.

Default interfaces somewhat change the equation: there's little risk with the proposed change (it's not breaking) _and_ you get performance benefits. Win win 🤷‍♂️

**Reply by ADD-David-Antolin** [reply]
Yes, of course, they are needed for not introducing breaking changes and that is fantastic, but for other classes that are not HttpRequestHeaders or HttpResponseHeaders you will not see any performance gains since the default implementations use the same indexer as before introducing those methods. As you stated, you will still see improvements in readability, usability and less errors in typing the name of headers wrong, which is also great.

**Comment by MauNguyenVan** [comment]
There is a bug related to partial meythod interface

Take a look at this code:

```cs
internal class DefaultInterfaceMemberBug
{
  public static void Main()
  {
    M(new C1()); // Print: 12 
    M(new C2()); // Print: ??, Can you guess?
  }

  static void M(I x)
  {
    x.M1();
    x.M2();
  }
}

partial interface I
{
  public void M1()
  {
    Console.Write(1);
  }

  public partial void M2();
  public partial void M2()
  {
    Console.Write(2);
  }
}

class C1 : I;

class C2 : I
{
  public void M1()
  {
    Console.Write(3);
  }

  public void M2()
  {
    Console.Write(4);
  }
}

```

Now, logically you might expect the output to be:
12  
34
But that's not what happens. Instead, the output is:
12  
32

This is not just limited to methods—it can happen with partial properties and partial indexers too.

This issue is being tracked here:
🔗 github.com/dotnet/roslyn/issues/77346

If you’re working with C# interfaces and using newer features like partial or default members, definitely check this out. It’s a subtle bug that could trip you up. 

. Net team will fix for partial property and partial even, but not for partial method, this fix will make inconsitancy behavior
if you want to Design team fix it, please go to github link to comment and vote it up
